---
title: Laboratory Task 2,3,4
---

## 01_Understanding Deep Learning

**Name:** Rosemarie Ann S. Bajao  
**Course:** BS Data Science  
**Activity:** Understanding Deep Learning  

---

## Overview

This notebook contains basic concepts in deep learning. It covers a simple neural network forward pass, error calculation, backpropagation, and a linear regression model implemented using PyTorch.

These sections include the concepts, calculations, code, and explanations needed to understand each task.

## Artificial Neural Network

An **Artificial Neural Network (ANN)** is a type of model made up of connected artificial neurons, or units, that are organized into layers. A typical ANN has an **input layer, one or more hidden layers, and an output layer**.

The input layer receives the data, while the hidden layers process and transform the information. The output layer then produces the final result. Connections between neurons have **weights**, which determine how much influence an input has on the calculation. **Biases** are also added during the calculations, and **activation functions** help determine the output of each neuron.

During training, the network uses examples from a dataset to learn patterns. Its output is compared with the expected result, and the error can be used to adjust the weights through **backpropagation**.


## Feedforward Neural Network

A **feedforward neural network** is one of the simplest types of artificial neural networks. In this type of network, information moves in **one direction**, starting from the input layer and passing through the hidden layers until it reaches the output layer.

The values are processed at each layer using weights, biases, and activation functions. This forward movement of information is called a **forward pass or forward propagation**.


## Forward Pass

A forward pass (also called **forward propagation**) is the process where input data flows through a neural network from the input layer to the output layer to generate a prediction.

### How It Works

* **Data Flow:** Data moves forward through the network, passing through the hidden layers until it reaches the output layer.
* **Calculations:** At each layer, neurons multiply the inputs by their weights, add a bias, and apply an activation function such as ReLU or Sigmoid.
* **No Learning:** During the forward pass, the weights and biases remain unchanged. The network only uses them to calculate the output.
* **Error Calculation:** After the output is produced, it can be compared with the target value to determine the error.

The results from the forward pass and loss calculation are then used during **backpropagation** to calculate the gradients. An optimizer can use these gradients to update the network's weights and biases.


# Laboratory Task 2

<div class="alert alert-block alert-success" style="font-family: Times New Roman">
    <h4><strong>Laboratory Task 2</strong></h4>

<p style="font-family:Times New Roman; text-align:justify; font-size:15px">
    <b>Instruction:</b> Perform a single forward pass and compute for the error.
</p>
</div>

$$x = \begin{bmatrix} 1 \\ 0 \\ 1 \end{bmatrix}$$

$$y = \begin{bmatrix} 1 \end{bmatrix}$$

$$f=max(0, Z_n)$$

$$\text{hidden unit weights} = 
\begin{bmatrix}
w_{11} = 0.2 && w_{12} = -0.3 \\
w_{13} = 0.4 && w_{14} = 0.1 \\
w_{15} = -0.5 && w_{16} = 0.2
\end{bmatrix}$$

$$\text{output unit weights} = 
\begin{bmatrix}
w_{21} = -0.3 \\
w_{22} = -0.2
\end{bmatrix}$$

$$\theta = 
\begin{bmatrix}
\theta_{1} = -0.4 \\
\theta_{2} = 0.2 \\
\theta_{3} = 0.1
\end{bmatrix}$$

In [1]:
import numpy as np

In [2]:
x = np.array([1, 0, 1], dtype=float)
y = np.array([1.0])

w_hidden = np.array([
    [0.2, -0.3],
    [0.4,  0.1],
    [-0.5, 0.2],
])
theta_hidden = np.array([-0.4, 0.2])

w_output = np.array([[-0.3], [-0.2]])
theta_output = np.array([0.1])

def relu(z):
    return np.maximum(0, z)

In [3]:
# Forward pass
Z_hidden = x.dot(w_hidden) + theta_hidden
H = relu(Z_hidden)
Z_output = H.dot(w_output) + theta_output
y_hat = relu(Z_output)

print(f"Z1 = {Z_hidden[0]:.4f}, Z2 = {Z_hidden[1]:.4f}")
print(f"H1 = {H[0]:.4f}, H2 = {H[1]:.4f}")
print(f"Z3 = {Z_output[0]:.4f}")
print(f"y_hat = {y_hat[0]:.4f}")

# Error
error = y_hat - y
sq_error = (y - y_hat)**2
print(f"\nError (y_hat - y) = {error[0]:.4f}")
print(f"Squared Error = {sq_error[0]:.4f}")

Z1 = -0.7000, Z2 = 0.1000
H1 = 0.0000, H2 = 0.1000
Z3 = 0.0800
y_hat = 0.0800

Error (y_hat - y) = -0.9200
Squared Error = 0.8464




The input values are first passed through the hidden layer using the given weights and biases. This produces Z1 = -0.7 and Z2 = 0.1. Applying the ReLU activation function changes Z1 to 0 while Z2 remains 0.1, giving H1 = 0 and H2 = 0.1.

The hidden-layer outputs are then passed to the output layer, producing Z3 = 0.08. Since the value is positive, ReLU keeps it as 0.08, which becomes the predicted output.

The target value is 1, while the prediction is only 0.08, so the prediction is lower than the target. Using the error calculation `ŷ - y`, the error is -0.92. The squared error is 0.8464.

## Backpropagation

**Backpropagation**, also called **backward propagation** or **backprop**, is a fundamental process used to train artificial neural networks. It calculates how the error affects the network's weights and biases through gradients, which are then used to update the parameters.

### How It Works

- **Forward Pass:** Input data moves through the network layers. Neurons use weights, biases, and activation functions to generate a prediction.
- **Error Calculation:** The predicted output is compared with the actual target value to determine the error.
- **Backward Pass:** The process starts at the output layer and moves backward through the network. It uses the **chain rule** to calculate how each weight and bias contributes to the error.
- **Parameter Update:** The gradients calculated during backpropagation are combined with the learning rate to update the weights and biases.

This process can be repeated during training, allowing the network to gradually improve its predictions.

# Laboratory Task 3

<div class="alert alert-block alert-success" style="font-family: Times New Roman">
    <h4><strong>Laboratory Task 3</strong></h4>

<p style="font-family:Times New Roman; text-align:justify; font-size:15px">
    <b>Instruction:</b> Perform a forward and backward propagation in python using the inputs from <b>Laboratory Task 2</b>
</p>

```python
x = np.array([1, 0, 1])
y = np.array([1])

# use relu as the activation function.

# learning rate
lr = 0.001
```
</div>

In [4]:
# Learning rate
lr = 0.001

# Derivative of ReLU
def relu_derivative(z):
    return (z > 0).astype(float)

In [5]:
# Forward Pass
Z_hidden = x.dot(w_hidden) + theta_hidden
H = relu(Z_hidden)

Z_output = H.dot(w_output) + theta_output
y_hat = relu(Z_output)

error = y_hat - y

print("Hidden weighted sum:", Z_hidden)
print("Hidden output:", H)
print("Output weighted sum:", Z_output)
print("Predicted output:", y_hat)
print("Error:", error)

Hidden weighted sum: [-0.7  0.1]
Hidden output: [0.  0.1]
Output weighted sum: [0.08]
Predicted output: [0.08]
Error: [-0.92]


In [6]:
# Backward Pass
delta_output = error * relu_derivative(Z_output)

grad_w_output = H.reshape(-1, 1) * delta_output
grad_theta_output = delta_output

delta_hidden = (
    w_output.flatten() * delta_output[0]
    * relu_derivative(Z_hidden)
)

grad_w_hidden = np.outer(x, delta_hidden)
grad_theta_hidden = delta_hidden

# Update Parameters
w_output = w_output - lr * grad_w_output
theta_output = theta_output - lr * grad_theta_output

w_hidden = w_hidden - lr * grad_w_hidden
theta_hidden = theta_hidden - lr * grad_theta_hidden

print("\nUpdated hidden weights:")
print(w_hidden)
print("\nUpdated output weights:")
print(w_output)
print("\nUpdated hidden biases:")
print(theta_hidden)
print("\nUpdated output bias:")
print(theta_output)


Updated hidden weights:
[[ 0.2      -0.300184]
 [ 0.4       0.1     ]
 [-0.5       0.199816]]

Updated output weights:
[[-0.3     ]
 [-0.199908]]

Updated hidden biases:
[-0.4       0.199816]

Updated output bias:
[0.10092]




The forward pass first produces the same values as the previous activity. The hidden-layer weighted sums are -0.7 and 0.1, which become 0 and 0.1 after applying ReLU. The output of the network is 0.08, while the target value is 1, resulting in an error of -0.92.

During backpropagation, the error is propagated backward through the output and hidden layers. The ReLU derivative is used to calculate the gradients for the network's weights and biases.

Because H1 was inactive after the ReLU function changed its value to 0, the parameters connected to H1 did not receive an update in this step. H2 was active, so the parameters connected to H2 were updated.

The calculated gradients are then used to update the weights and biases using the learning rate of 0.001. The updated parameters are slightly different from their original values. These updates are intended to move the network's prediction closer to the target value in the next forward pass.

## Introduction to PyTorch

**PyTorch** is an open-source framework used for building and training machine learning and deep learning models. In this activity, PyTorch is used to implement and train a **linear regression model**.

### PyTorch Components

For the linear regression example, the following PyTorch components are used:

- **Input and Target:** The input data and expected target values are defined for the model.
- **Dataset and DataLoader:** The dataset stores the samples and their corresponding targets, while the DataLoader provides the data in batches during training.
- **`nn.Linear`:** A fully connected layer used to connect the input and output features of the model. Two `nn.Linear` layers are used in the regression model.
- **Loss Function:** Measures the difference between the model's predictions and the target values.
- **Optimizer:** Uses the calculated gradients to update the model's parameters.
- **Training:** The model repeatedly processes the data, calculates the loss, and updates its parameters to improve its predictions.

## Linear Regression Using PyTorch

Linear regression is used to model the relationship between input variables and continuous output values. In this activity, PyTorch is used to build and train a simple linear regression model.

The dataset contains three input variables: **temperature, rainfall, and humidity**. The model uses these values to predict two target values: **apples and oranges**.

The model is trained using **Mean Squared Error (MSE)** as the loss function and **Stochastic Gradient Descent (SGD)** as the optimizer.

<div class="alert alert-block alert-success" style="font-family: Times New Roman">
    <h4><strong>Laboratory Task 4</strong></h4>

<p style="font-family:Times New Roman; text-align:justify; font-size:15px">
    <b>Instruction:</b> Train a linear regression model in PyTorch using a regression dataset. Use the following parameters.
</p>

<ul>
    <li>Criterion: MSE Loss</li>
    <li>Fully Connected Layers x 2</li>
    <li>Batch Size: 8</li>
    <li>Optimizer: SGD</li>
    <li>Epoch: 1000</li>
</ul>
</div>

In [7]:
# pip install torch

In [8]:
import torch
import torch.nn as nn
from torch.utils.data import TensorDataset, DataLoader

In [9]:
# Dataset
x_train = np.array([
    [73, 67, 43], [91, 88, 64], [87, 134, 58],
    [102, 43, 37], [69, 96, 70], [73, 67, 43],
    [91, 88, 64], [87, 134, 58], [102, 43, 37],
    [69, 96, 70], [73, 67, 43], [91, 88, 64],
    [87, 134, 58], [102, 43, 37], [69, 96, 70]
], dtype='float32')

y_train = np.array([
    [56, 70], [81, 101], [119, 133],
    [22, 37], [103, 119], [56, 70],
    [81, 101], [119, 133], [22, 37],
    [103, 119], [56, 70], [81, 101],
    [119, 133], [22, 37], [103, 119]
], dtype='float32')

In [10]:
# Convert data to PyTorch tensors
inputs = torch.from_numpy(x_train)
targets = torch.from_numpy(y_train)

# Create dataset and DataLoader
train_ds = TensorDataset(inputs, targets)

batch_size = 8
train_dl = DataLoader(train_ds, batch_size=batch_size, shuffle=True)

In [11]:
# Define the model
model = nn.Sequential(
    nn.Linear(3, 8),
    nn.Linear(8, 2)
)

In [12]:
# Loss function and optimizer
criterion = nn.MSELoss()
optimizer = torch.optim.SGD(model.parameters(), lr=0.00001)

In [13]:
# Training
num_epochs = 1000

for epoch in range(num_epochs):
    for xb, yb in train_dl:

        # Forward pass
        predictions = model(xb)

        # Calculate loss
        loss = criterion(predictions, yb)

        # Backward pass
        optimizer.zero_grad()
        loss.backward()

        # Update parameters
        optimizer.step()

    if (epoch + 1) % 100 == 0:
        print(f"Epoch [{epoch + 1}/{num_epochs}], Loss: {loss.item():.4f}")

Epoch [100/1000], Loss: 21.9726
Epoch [200/1000], Loss: 4.9516
Epoch [300/1000], Loss: 2.0179
Epoch [400/1000], Loss: 1.0537
Epoch [500/1000], Loss: 0.9554
Epoch [600/1000], Loss: 0.7864
Epoch [700/1000], Loss: 0.6783
Epoch [800/1000], Loss: 0.8528
Epoch [900/1000], Loss: 0.5504
Epoch [1000/1000], Loss: 0.5841


In [14]:
with torch.no_grad():
    predictions = model(inputs)
    final_loss = criterion(predictions, targets)

print("\nFinal MSE Loss:", final_loss.item())

print("\nPredicted values:")
print(predictions)

print("\nActual values:")
print(targets)


Final MSE Loss: 0.518781304359436

Predicted values:
tensor([[ 57.0741,  70.2041],
        [ 82.3257, 100.6295],
        [118.4909, 133.2311],
        [ 21.0569,  37.0635],
        [102.0608, 118.9028],
        [ 57.0741,  70.2041],
        [ 82.3257, 100.6295],
        [118.4909, 133.2311],
        [ 21.0569,  37.0635],
        [102.0608, 118.9028],
        [ 57.0741,  70.2041],
        [ 82.3257, 100.6295],
        [118.4909, 133.2311],
        [ 21.0569,  37.0635],
        [102.0608, 118.9028]])

Actual values:
tensor([[ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.],
        [ 56.,  70.],
        [ 81., 101.],
        [119., 133.],
        [ 22.,  37.],
        [103., 119.]])




The model was trained using the given regression dataset, where temperature, rainfall, and humidity were used as the input variables and apples and oranges were used as the target values.

The model uses two fully connected layers. The first layer takes the three input features and produces eight intermediate values, while the second layer uses those eight values to produce the two target outputs. During each training step, the model generates predictions, calculates the MSE loss, performs backpropagation, and updates its parameters using SGD.

The training was performed for 1000 epochs with a batch size of 8. The overall decrease in loss during training indicates that the model was adjusting its parameters to make predictions closer to the target values.

The final MSE loss represents the average squared difference between the model's predictions and the actual target values. A lower MSE indicates that the predictions are closer to the actual values.

## Conclusion

The different sections connect to show how a neural network processes data, calculates errors, updates its parameters, and is eventually trained using a framework such as PyTorch.

The forward pass showed how the input values move through the hidden and output layers to produce a prediction. The input values were multiplied by the given weights and combined with the biases to calculate the hidden-layer values. The ReLU activation function then changed the negative value of -0.7 to 0, while the positive value of 0.1 remained unchanged. This resulted in H1 = 0 and H2 = 0.1. The hidden-layer outputs were then passed to the output layer, producing a predicted value of 0.08. Compared with the target value of 1, the prediction resulted in an error of -0.92 and a squared error of 0.8464.

The forward pass also showed how an activation function can affect the rest of the training process. Since the value for H1 was negative before applying ReLU, it became 0 and did not contribute to the output in that forward pass. During backpropagation, the error was passed backward through the network and gradients were calculated using the ReLU derivative. Because H1 was inactive, the parameters connected to it did not receive an update in that step, while the parameters connected to the active hidden unit H2 were updated. The learning rate of 0.001 was then used when updating the weights and biases.

The PyTorch section showed how the same general process can be applied using a framework. The regression model used temperature, rainfall, and humidity as the three input features and predicted apples and oranges as the two target values. The model used two fully connected layers, with the first layer changing the three input features into eight intermediate values and the second layer producing the two outputs. MSE Loss was used to measure the prediction error, while SGD was used to update the model parameters. The model was trained using a batch size of 8 for 1000 epochs.

The choice of learning rate also had a clear effect on the training result. A learning rate of 0.0001 caused the loss to become `nan`, while the smaller learning rate of 0.00001 allowed the model to train more stably and reach a final MSE loss of about 0.5. This shows that the learning rate controls how much the model's parameters change during each update, so the value used during training can affect whether the model is able to learn properly.

Overall, the concepts of artificial neural networks, forward propagation, error calculation, backpropagation, and PyTorch training are connected through the same basic learning process. The manual calculations made it possible to see what happens inside the network step by step, while the PyTorch implementation showed how many of these calculations can be handled automatically when training a model.